AppDelegate.swift

```diff
@@ -0,0 +1,122 @@
+//
+//  AppDelegate.swift
+//  srushtika-tut-fixes
+//
+//  Created by Srushtika Neelakantam on 06/01/2020.
+//  Copyright © 2020 Srushtika Neelakantam. All rights reserved.
+//
+
+import UIKit
+import Ably
+import UserNotifications
+
+let authURL = ""
+let API_KEY = ""
+let myClientId = "abc"
+let ablyClientOptions = ARTClientOptions()
+
+
+@UIApplicationMain
+class AppDelegate: UIResponder, UIApplicationDelegate, ARTPushRegistererDelegate {
+    func didDeactivateAblyPush(_ error: ARTErrorInfo?) {
+        print("** push deactivated")
+    }
+    
+
+    var window: UIWindow?
+    var realtime: ARTRealtime!
+    var channel: ARTRealtimeChannel!
+    var subscribed = false
+    
+    //on launch, init ably and call activate
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        // Override point for customization after application launch.
+        
+        print("** app launched")
+
+       UNUserNotificationCenter.current().requestAuthorization(options:[.badge, .alert, .sound]) { (granted, err) in
+           DispatchQueue.main.async() {
+               UIApplication.shared.registerForRemoteNotifications()
+               print("** request to show notifications successful")
+           }
+       }
+        
+       self.realtime = self.getAblyRealtime()
+       self.realtime.push.activate()
+        
+        return true
+    }
+    
+    //register with APNs - returns deviceToken
+    func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
+        DispatchQueue.main.async() {
+            print("** didRegisterForRemoteNotificationsWithDeviceToken")
+            ARTPush.didRegisterForRemoteNotifications(withDeviceToken: deviceToken, realtime: self.getAblyRealtime())
+        }
+    }
+
+    //register with APNs with error - returns error
+    func application(_ application: UIApplication, didFailToRegisterForRemoteNotificationsWithError error: Error) {
+        DispatchQueue.main.async() {
+            print("** didFailToRegisterForRemoteNotificationsWithError")
+            ARTPush.didFailToRegisterForRemoteNotificationsWithError(error, realtime: self.getAblyRealtime())
+        }
+    }
+    
+    // function to init ably with client options - key and clientId
+    private func getAblyRealtime() -> ARTRealtime {
```

Comment:
I would call this `initAblyRealtime` instead of `getAblyRealtime`. The method is creating a new Realtime instance every time it's called, so it can be misleading for the user and abuse it.

Reply:
> The method is creating a new Realtime instance every time it's called

Which it really shouldn't. The instance that `push.activate` is called on must be the same instance that is passed to these two methods.

`getAblyRealtime` should initialize `self.realtime` or return it if it's already initialized.